### ISSUE TITLE
Unclear on the use the `jx` module

### ISSUE THREAD

**Issue by Adeynack** [SalomonBrys/Kodein#66]
I am have bound services needing other bound services.

My binding looks like this:
```kotlin
val kodein = Kodein {
    import(jxInjectorModule)
    bind<Properties>() with singleton { PropertiesLoaderUtils.loadProperties(ClassPathResource("application.properties")) }
    bind<FileContentProvider>() with singleton { FileContentProvider.default }
    bind<DistributionService>() with singleton { jx.newInstance<DistributionServiceImpl>() }
    bind<InnkeeperClient>() with singleton { jx.newInstance<InnkeeperClientImpl>() }
    bind<QuiltClient>() with singleton { jx.newInstance<QuiltClientImpl>() }
    bind<TessellateClient>() with singleton { jx.newInstance<TessellateClientImpl>() }
    bind<JsonService>() with singleton { JsonServiceImpl() }
    bind<RequestContext>() with singleton { RequestContextImpl() }
    bind<Authorization>() with singleton { AuthorizationImpl() }
}
```

The moment I hit the resolving of the `DistributionService`, it fails with this message:

```
com.github.salomonbrys.kodein.Kodein$DependencyLoopException: Dependency recursion:
       ╔═> bind<JxInjector>()
       ╠─> bind<DistributionService>()
       ╚═> bind<JxInjector>()
```

The constructor of my `DistributionServiceImpl` is:
```kotlin
class DistributionServiceImpl(
    private val jsonService: JsonService,
    private val properties: Properties,
    private val quiltClient: QuiltClient,
    private val innkeeperClient: InnkeeperClient,
    private val tessellateClient: TessellateClient
) : DistributionService {
```

When I debug with breakpoints in every one of the `singleton` lambda, I observe that the following are called:
- `JsonServiceImpl()`
- `FileContentProvider.default`
- `jx.newInstance<DistributionServiceImpl>()`
Then it fails before entering the constructor of class `DistributionServiceImpl`.

I feel that I simply am not using this right. However, the documentation does not state a clear example of how to declare a binding block using the `jx` approach. It just states as an example:

```
val kodein = Kodein {
    import(jxInjectorModule)
    /* Other bindings */
}
```

I am interested in HOW to declare the _other bindings_ here.

Thanks for whatever help you can provide.

**Comment by SalomonBrys** [comment]
Hi there,
this really looks like a bug :(
I;ll have a look as soon as I can

**Comment by Adeynack** [comment]
But @SalomonBrys , could you please first confirm that this is the expected _recipe_ in the `jx` approach?

```kotlin
val kodein = Kodein {
    import(jxInjectorModule)
    bind<FooInterface>() with singleton { jx.newInstance<FooImplementation>() }
    bind<BarInterface>() with singleton { jx.newInstance<BarImplementation>() }
}
```

**Comment by SalomonBrys** [comment]
Certainly looks correct...

**Comment by SalomonBrys** [comment]
That was indeed a bug on the JxInject part.
It happend especially when using this construct (having dependencies with transient dependencies, all created via `jx.newInstance`.

The fix will be deployed with version 4.1.

Please note that, while this will work in 4.1, this construct is very much discouraged as it uses a lot more reflection than the standard construct.
Please consider using the standard construct, using `instance()` for each transient dependencies as it is much more kotlin idiomatic and optimised.

```kotlin
bind<FooInterface>() with singleton { FooImplementation(instance(), instance()) }
```

Note that you can also make `FooImplementation` implement `KodeinAware` so that it can use delegation to get it's dependencies (rather than from its constructor).

```kotlin
bind<FooInterface>() with singleton { FooImplementation(kodein()) }
```

**Comment by Adeynack** [comment]
We are exploring possibilities and have, with Kodein, 3 branches (to assert what we like the most):
1. Kodein as the only parameter of all dependencies (then, private `val Foo: IFoo = kodein.instance() inside`)
1. Normal constructor parameters and using the `bind<IFoo> with singleton { FooImpl(instance(), instance() ,instance() }` at wiring time
1. Normal constructor parameters using the `jx` extension version

I think so far that option 2 is the most pragmatic and does not require reflection nor extensions.

FYI, we are also playing with:
- Guice, which looks great in Scala with the way _Play!_ wrapped it up, but looks horrible in Kotlin without any proper DSL. 😩
- Spring, that supposedly has a nice Kotlin DSL (haven't tried yet)
